WordPress/gutenberg
# [RNMobile] Resolve Text Entry Issues/Errors in Alt Text Settings

**Fixes:** 

* https://github.com/WordPress/gutenberg/issues/33857
* https://github.com/WordPress/gutenberg/issues/33856
* https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368

**Gutenberg Mobile:**

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/3818

## Description

The code introduced in this PR addresses the following three issues with the text entry field in the image block's alt text settings:

* The cursor would jump to different positions at times when typing swiftly or deleting text on Android.
* Text would duplicate itself when typing swiftly in both Android and iOS.
* The following error was logged in the developer console when opening the image block's settings:

```
ExceptionsManager.js:180 Warning: Cannot update a component (`Unknown`) while rendering a different component (`BottomSheetTextControl`). To locate the bad setState() call inside `BottomSheetTextControl`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
    in BottomSheetTextControl (at edit.native.js:416)
```

An open issue in the React Native repository, https://github.com/facebook/react-native/issues/30503, lines up with the behavior described in the first two issues. That issue outlines the fact that this was introduced in React Native 0.63.0 and is related to global state management with the `TextInput` component. The third issue is semi-related, and resolved through the code from this PR, given it also touches on state management.

Further technical details around the code changes are provided in the **Types of changes** section below.

## How has this been tested?

I completed the following test cases on a physical **Pixel 5, Android version 11** and **an iPhone XR, iOS 14.6**. 

While going through the below flows, it'd be useful to also think about and experiment with different edge cases e.g. entering return at random while typing swiftly or clicking delete abruptly.

### Test Case 1: General Flow

<details>
<summary>To start with, the general flow for adding new alt text is as follows.</summary>

* In the app, navigate to _Home_ → _Blog Posts_ to create a new blog post or open an existing one. Add a new image block and upload an image to it.
* Tap the image block's cog/gear icon to access its settings. 
* Tap **Alt Text** to access the text editor for adding alt text to the image.
* Enter new text into the text editor and exit the settings.
* Confirm that the image's alt text has been correctly updated. This can be further verified by saving changes and confirming the alt text has been correctly updated on the web.

</details>

### Test Case 2: Deleting Text

<details>
<summary>We should also verify that the cursor no longer changes position at random after deleting text.</summary>

* Follow the first three steps in the **General Flow** test case to access the alt text settings within an image block's settings panel.
* After entering text into the editor, move the cursor somewhere in the middle and press the delete/backspace key.
* Verify that one character is deleted and the cursor doesn't change position. 
</details>

### Test Case 3: Typing Swiftly

<details>
<summary>Next, we should verify that it's possible to type swiftly into the text editor, with no unwanted side effects, such as duplication.</summary>

* Again, follow the first three steps in the **General Flow** test case to access the alt text settings within an image block's settings panel.
* Attempt to enter a sentence describing the image very quickly into the editor. 
* Verify that the text you've entered appears correctly in the editor, with no duplication or errors. 
</details>

### [Developer] Test Case 4: Verify No Errors in Developer Console

While testing the branch on a local setup, confirm that there is no longer a warning related to the `BottomSheetTextControl` component when opening the image block's settings.

## Screenshots

Previously, hitting the backspace/delete key would lead to the cursor jumping, but you'll see this is no longer the case in the following video:

https://user-images.githubusercontent.com/2998162/128915167-4389a02c-94a5-4292-871d-b07f2968a607.mov

## Types of changes

This PR introduces a bug fix (a non-breaking change which fixes an issue) with the following high-level overview of the code changes:

* React Native's `TextInput` component is being used in [the `BottomSheetTextControl` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet-text-control/index.native.js). It's this component that's used to enable users to enter alt text via the image block's settings.
* The `onChange` prop was being used to create a side effect when text was typed into the `TextInput` component. As [`onChange` uses `setState`](https://github.com/facebook/react/issues/18178#issuecomment-596880684), this was creating [the error outlined here](https://github.com/WordPress/gutenberg/pull/33057#pullrequestreview-725305368). The solution was to [remove the reference to `onChange`](https://github.com/WordPress/gutenberg/pull/33975/files#diff-13d6976d9626e0d1ff8c730f322d911b37ce33194623cd7d14b11ae146d5636eL81) and [move the functionality to a `useEffect` hook](https://github.com/WordPress/gutenberg/pull/33975/files#diff-13d6976d9626e0d1ff8c730f322d911b37ce33194623cd7d14b11ae146d5636eR47).
* In addition to the above changes, the `TextInput` component's `value` prop was [changed to `defaultValue`](https://github.com/WordPress/gutenberg/pull/33975/files#diff-13d6976d9626e0d1ff8c730f322d911b37ce33194623cd7d14b11ae146d5636eR85), as per [the suggestion in this open bug report](https://github.com/facebook/react-native/issues/30503#issuecomment-787001771) to workaround a known React Native issue with the `TextInput` component. 

<hr />

## Checklist:
